Feature/quark smearing.pr

alexstrel opened the pull request:
PR introduces an implementation of the Gaussian smearing for staggered quarks. It also includes computation of two link field, simple test instruction: 
 mpirun -np 1 ./tests/staggered_gsmear_test --dslash-type asqtad --prec double --gridsize 1 1 1 1 --dim 16 16 16 48 --load-gauge path-togauge-config

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Add heatbath example

cpviolator committed:
Tidy up

cpviolator committed:
Fix bug in heatbath example

cpviolator committed:
Add checkpointing to heatbath

cpviolator committed:
Updated examples README

cpviolator committed:
Remove stray git merge marker

cpviolator committed:
Fix hardcoded temporal dim instantiation, clean up reporting data

Commit:
fix a bug in the naming of output correlators when using multiple gpu

cpviolator committed:
Remove Propgator class.... wait for multi RHS framework

cpviolator committed:
Merge branch 'feature/hotqcd_correlators_gk_merged' of https://github.com/lattice/quda into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Remove duplicated code from tests, migrate to the examples

cpviolator committed:
Move correlator enums and params to example file, adda parse option for masses and kappa, hardcode QUDA clover construction and detail how to retrieve the data in a comment

cpviolator committed:
Use an array for multiple kappa/mass value input

cpviolator committed:
Use kappa arrays, shift all operator construction to helper

cpviolator committed:
Add a csw term option

cpviolator committed:
Deliniate between Csw and clover_coeff, preserve behaviour where clover_coeff is set to Csw * kappa by the user

cpviolator committed:
Gaussian fermion smearing

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Remove propagator.h header

cpviolator committed:
Remove propagator code from examples

Commit:
a minor change in the naming for the channel of output correlators

cpviolator committed:
GK merge

cpviolator committed:
Add the option to smear the gauge field prior to inversion

cpviolator committed:
Fix profiling and stdout

cpviolator committed:
Clarify a command line param

cpviolator committed:
Fix MPI stdout, add a gaussian smearing operator

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
GK merge

Commit:
Generalized contraction code to include nSpin=1 (staggered) as well as nSpin=4.

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
GK merge

Commit:
Added staggered meson contraction with Fourier phase summed over spatial volume.

Commit:
Merge branch 'feature/hotqcd_correlators_gk_merged' into feature/hotqcd_correlators_gk_merged_JNS

Commit:
Modifications to allow specifying either a cosine or sine Fourier transform in meson contractions instead of an exponential phase.

cpviolator committed:
Clean up prior to GK merge

cpviolator committed:
Fixes for externals

cpviolator committed:
GK merge, adjust the chiral projector template

detar committed:
Fix typos

detar committed:
Add definitions: qudaContractFT and QudaContractArgs_t

detar committed:
Fix contraction-type string

detar committed:
Fix typo

detar committed:
Add qudaContractFT

detar committed:
Add contractFT_test

detar committed:
Set cs_param values explicitly for transparency.  Temporarily restrict to staggered contractions.

cpviolator committed:
Correct the help output for source/sink smearing

luhuhis committed:
temporarily add correlator_test.cpp for more convenient development (instead of correlator_quda.cpp). add first steps for routine to convert 4D sources to 5D for MDWF.

cpviolator committed:
Mid Debug SUMMIT -> LASSEN

cpviolator committed:
Remove debug statements, correct the array sizes, fix various warnings

Commit:
fixed compiling errors

luhuhis committed:
add minor comments

luhuhis committed:
fix small merge conflict

luhuhis committed:
complete conversion routine from 4D to 5D point source. needs testing

luhuhis committed:
start fixing bugs

luhuhis committed:
apply more fixes for the 5D MDWF correlator computation

luhuhis committed:
clarify comments

luhuhis committed:
move fields to device in 4D to 5D conversion

luhuhis committed:
identify cause of segmentation fault. no fix yet

luhuhis committed:
change nDim=5 to nDim=4 for colorspinorfields

luhuhis committed:
more fixes trying to solve the 4D to 5D conversion

luhuhis committed:
fix formal bugs in convert4Dto5DpointSource

luhuhis committed:
try to fix empty 5D source

cpviolator committed:
minor change in help output

cpviolator committed:
Fix Conflicts

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Add precision to the correlation functions for comparison testing

cpviolator committed:
GK merge

cpviolator committed:
GK merge

cpviolator committed:
Merge branch 'feature/hotqcd_correlators_gk_merged' of https://github.com/lattice/quda into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

cpviolator committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/hotqcd_correlators_gk_merged

alexstrel committed:
Merge branch 'feature/hotqcd_correlators_gk_merged' into feature/staggered_correlators_gk

alexstrel committed:
Merge branch 'feature/generic_kernel' into feature/staggered_correlators_gk

alexstrel committed:
a few bugs fixed

alexstrel committed:
another bug fix (explicite PC)

alexstrel committed:
fixed contraction number in the interface routine

alexstrel committed:
Merge branch 'feature/generic_kernel' into feature/staggered_correlators_gk

alexstrel committed:
some memory optimization

Commit:
Fourier phase computation reproduces milc in standalone C++ testing. Added more momenta cases to contraction test code.

Commit:
Merge branch 'feature/staggered_correlators_gk' of https://github.com/lattice/quda into feature/staggered_correlators_gk

alexstrel committed:
replaced constexpr if statement

Commit:
Merge remote-tracking branch 'refs/remotes/origin/feature/staggered_correlators_gk' into feature/staggered_correlators_gk

Commit:
In the milc contraction interface, csParam.pc_type = QUDA_4D_PC.

Commit:
Change momenta used in contractFT_test.

Commit:
Change output fromat of contraction results.

Commit:
Add momentum test cases.

Commit:
Not thread safe: return of pointer to static variable within a function

Commit:
Fixed SpinColor field mapping for milc contractions.

alexstrel committed:
Merge branch 'feature/generic_kernel' into feature/quark-smearing

alexstrel committed:
changed class names for staggered qsmear

alexstrel committed:
minor changes

alexstrel committed:
added 2 link computations, corrected quark smearing kernels

alexstrel committed:
fixed typo

Commit:
Merge branch 'feature/generic_kernel' into feature/staggered_correlators_gk

alexstrel committed:
Merge branch 'feature/generic_kernel' into feature/quark-smearing

alexstrel committed:
updated files with content

alexstrel committed:
fixed compiler errors

alexstrel committed:
fixed missing forward index

alexstrel committed:
Merge branch 'feature/generic_kernel' into feature/staggered_correlators_gk

alexstrel committed:
Merge branch 'feature/staggered_correlators_gk' into feature/quark-smearing

alexstrel committed:
corrected previous merge

alexstrel committed:
fixed compilation errorsx

alexstrel committed:
Merge branch 'feature/quark-smearing' of https://github.com/lattice/quda into feature/quark-smearing

alexstrel committed:
fixed previous commit, fixed linker errors when contraction is not enabled

alexstrel committed:
removed redefined var

alexstrel committed:
fixed previous commits (compiler error from test applications)

alexstrel committed:
removed artifacts

alexstrel committed:
Merge branch 'feature/staggered_correlators_gk' into feature/quark-smearing

alexstrel committed:
added interface routine

alexstrel committed:
corrected halo contribution

alexstrel committed:
fixed low-level interface methods

alexstrel committed:
Merge branch 'feature/staggered_correlators_gk' into feature/quark-smearing

alexstrel committed:
Merge branch 'feature/generic_kernel' into feature/quark-smearing

alexstrel committed:
compilation of the laplace dslash is disabled, multi-gpu init is
temporarily disabled

alexstrel committed:
a few fixes for the previous GK branch merge (not tested yet!)

alexstrel committed:
minor cleaning

alexstrel committed:
a few corrections

alexstrel committed:
redesigned interface fucntions, added missing method for staggered
operator

alexstrel committed:
checked interface implmentation, added a placeholder for milc interface

alexstrel committed:
a minor correction in the interface args

alexstrel committed:
Merge branch 'feature/generic_kernel' into feature/quark-smearing

alexstrel committed:
added test program, cleaned compute kernel interface

alexstrel committed:
Merge branch 'feature/generic_kernel' into feature/quark-smearing

hwancheolJeong committed:
Fix staggered two-link smearing routines.

alexstrel committed:
2 link compute stage is now isolated in a separate files

alexstrel committed:
Merge branch 'feature/generic_kernel' into feature/quark-smearing

alexstrel committed:
added explicit allocation of the extended field for two-link
computations

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing

alexstrel committed:
using extended field, still get segfault

alexstrel committed:
fixed segfault

alexstrel committed:
minor cleaning

hwancheolJeong committed:
fix minor cleaning

alexstrel committed:
fixed multi-gpu version

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing

alexstrel committed:
reformulated smearing operator for the staggered case

alexstrel committed:
minor fixes

alexstrel committed:
updated SmearOp method

alexstrel committed:
removed artifacts, fixed gauge field argument

hwancheolJeong committed:
fix a typo in CMake

hwancheolJeong committed:
fix a missing input parameter

hwancheolJeong committed:
fix an input parameter order

hwancheolJeong committed:
fix two-link computation

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing

alexstrel committed:
Added internal two-link test

alexstrel committed:
updated version, with corrections. Do we need another routine with explicit input gauge field?

alexstrel committed:
now allowed external input links, still not tested

alexstrel committed:
fixed compilation errors

alexstrel committed:
a few bugs fixed. Gaussian smearing is now computed via extended fields (as it should). Two link profiling might be incomplete.

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing

alexstrel committed:
a few changes, minor cleaning

hwancheolJeong committed:
Fix CMake for staggered_twolink_test

hwancheolJeong committed:
Fix quark smearing kernel and two-link smearing interface

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing

alexstrel committed:
corrected two-link computing interface function

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing

alexstrel committed:
initial bunch of changes to get t0 functionality

alexstrel committed:
modified compute kernels for t0 bulk computing

alexstrel committed:
fixed compiler errors, changed interface function

alexstrel committed:
fixed cmake file

alexstrel committed:
synchronized with the local branch

alexstrel committed:
Merge branch 'feature/quark-smearing' of https://github.com/lattice/quda into feature/quark-smearing

alexstrel committed:
forgot to update file for the previous update

alexstrel committed:
a temporary fix multi-gpu version, not performant (coming next)

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing

hwancheolJeong committed:
fix t0 parameter implementation to work with multi nodes

alexstrel committed:
fixed autotuning issue

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing

alexstrel committed:
swapped boolean values

hwancheolJeong committed:
t0 parameter supports splitting in time direction.

hwancheolJeong committed:
Disable normalizing source after smearing

hwancheolJeong committed:
Prohibit local lattice size smaller than 6 for partitioned directions

hwancheolJeong committed:
Clean-up and minor fixes

hwancheolJeong committed:
Add two-link Gaussian smearing flag, and fix flops

hwancheolJeong committed:
Smearing single parity site field invokes error.

hwancheolJeong committed:
Divide two-link kernel into four direction parts

hwancheolJeong committed:
Merge branch 'develop' up to commit 470cf533d9980333297ca3442685d69ce2fd9ed8

hwancheolJeong committed:
Clean up updates irrelevant to the two-link gaussian quark smearing

hwancheolJeong committed:
Add GPU_TWOLINK_GSMEAR flag requirement to StaggeredQSmear.

hwancheolJeong committed:
Merge branch 'develop' into feature/quark-smearing.merge

* develop: (238 commits)
  Fix a bug in `invert_x_batch.h` where the coefficient size is not equal to the input vector size for `axpy`. Added a corresponding invert ctest parameter dimension to catch this.
  Display warning if second-stage tuning is slower than first stage.  Added a warm up call prior to hot loop to minimize false warnings
  Use std::vector over new/delete.  Fix aux printing while tuning
  Fix bug in VUV kernel: must set arg.coarse_color_wave in apply depending on tp.aux.w
  Made the include of json.hpp come from external
  Half precision staggered fields now use a single 128-bit word per site, composed of 6 x short and float
  Fix bug in hotDotProduct unit test, which can cause test failures
  Set missing location in multigrid_evolve_test
  Fix multiReduce_recurse: prior identification of lower trinagular blocks was buggy for non-powers-of-two sizes.  This fixes hDotProduce.  Add a unit test for hDotProduct
  Cleanup to multi-blas
  Fix some hip typos
  Apply clang-format
  fix typo
  Add missing 96/96 instantions, and fix ifdef bug for 24/24 inclusion
  Fix clang compilation issue (use class name, not variable name, when addressing a static class member)
  Fix clang compilation
  Fix some false warning pre-CUDA 11.5
  include multigrid.h for instantiatePrecisionMG
  Fix typo in HIP target math_helper.cuh
  Update QIO tag to 3.0
  ...

hwancheolJeong committed:
Fix lattice dimension type

hwancheolJeong committed:
Add parameter for deleting two-link after smearing

hwancheolJeong committed:
Add function freeing gaugeSmeared (or two-link)

hwancheolJeong committed:
Fix initial kernel dimension

hwancheolJeong committed:
Allow gauge field reconstruction for two-link computation

hwancheolJeong committed:
Merge branch 'develop' into feature/quark-smearing.pr

alexstrel committed:
added test applications

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

alexstrel committed:
an update with gsmear test application

alexstrel committed:
added missing files

alexstrel committed:
minor cleaning , removed artifacts

mathiaswagner commented on lib/CMakeLists.txt:
```diff
@@ -352,6 +352,10 @@ if(QUDA_LAPLACE)
   target_compile_definitions(quda PUBLIC GPU_LAPLACE)
 endif(QUDA_LAPLACE)
 
+if(QUDA_TWOLINK_GSMEAR)
```
is that used anywhere or can we just remove the `if` ?

hwancheolJeong replied:
Sorry for the slow response. We intend for users to turn on this flag when building QUDA only if they need it. We will create a wiki page for it.

mathiaswagner replied:
I was more worried that there is no corresponding cmake option.

hwancheolJeong replied:
Thanks for letting me know that! I have just added the cmake option.

mathiaswagner commented on tests/utils/command_line_params.cpp:
```diff
@@ -1023,3 +1029,13 @@ void add_testing_option_group(std::shared_ptr<QUDAApp> quda_app)
+{
+  auto opgroup = quda_app->add_option_group("Quark smearing", "Options controlling quark smearing testing");
+  opgroup->add_option("--compute_two_link", compute_two_link, "Compute two link field (default true)");
+  opgroup->add_option("--delete_two_link", delete_two_link, "Compute two link field (default true)");
```
the help text needs an update

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

hwancheolJeong committed:
Add cmake option for QUDA_TWOLINK_GSMEAR flag

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

alexstrel committed:
minor corrections

alexstrel committed:
reordered body of the two-link function, to reconcile with jenkins

alexstrel committed:
more fixes required from jenkins

alexstrel committed:
cleaned unused vars

alexstrel committed:
removed unused vars

alexstrel committed:
one more unsued var deleted

alexstrel committed:
fixed compiler warnings/errors for non-multigpu stuff

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr
a few corrections in CL help information

maddyscientist commented on include/dirac_quda.h:
```diff
@@ -230,6 +230,16 @@ namespace quda {
+    virtual void SmearOp(ColorSpinorField &, const ColorSpinorField &, 
+                         const double&, const double &, const int &,  const QudaParity) const 
+    {
+      errorQuda("Not implemented!\n");
```
We don't need the trailing `\n` for `errorQuda` calls since it will apply this automatically.

alexstrel replied:
Done.

maddyscientist commented on include/dirac_quda.h:
```diff
@@ -1326,6 +1336,8 @@ namespace quda {
     void createCoarseOp(GaugeField &Y, GaugeField &X, const Transfer &T, double kappa, double mass, double mu = 0.,
                         double mu_factor = 0., bool allow_truncation = false) const;
+
+    void SmearOp(ColorSpinorField &out, const ColorSpinorField &in, const double &a, const double &b, const int &t0,  const QudaParity parity) const;
```
doxygen for this function

alexstrel replied:
done

maddyscientist commented on include/dirac_quda.h:
```diff
@@ -1556,6 +1568,8 @@ namespace quda {
     */
     virtual void prefetch(QudaFieldLocation mem_space, qudaStream_t stream = device::get_default_stream()) const;
+    
+    void SmearOp(ColorSpinorField &out, const ColorSpinorField &in, const double &a, const double &b, const int &t0,  const QudaParity parity) const;    
```
doxygen for this function

alexstrel replied:
done.

maddyscientist replied:
Thanks, but there's still some missing `in`/`out` attributes and the ones for `in` and `out` should be `@param[out] out` as opposed to `@param out[out]`

alexstrel replied:
fixed

mathiaswagner replied:
Why do we have a `b (ignored)` parameter?

mathiaswagner replied:
Why do we pass doubles as const ref instead of by value?

hwancheolJeong replied:
In fact, both `a` and `b` parameters do nothing in `Dirac(Improved)Staggered::SmearOP()`'s, since we pull them out as an overall factor. (For clarity, we omit their names in the definitions of these functions.)
However, we keep them in the input structure to accommodate the future addition of `GaugeLaplace::SmearOp()`.
 (See https://github.com/lattice/quda/blob/feature/quark-smearing/include/dirac_quda.h#L1959. I suppose Dean worked on it before us.)
If you want to remove them for now, that is okay with us.

hwancheolJeong replied:
Regarding calling by reference, again, we follow the structure of `GaugeLaplace::SmearOp()` that started before this work. I don't think we have any preference for it.

hwancheolJeong replied:
Okay. Alexei and I discussed it, and we decided to change these `a`,`b`,`t0` parameters to be called by value and fixed doxygen description for `a`. Thank you for bringing it up.
https://github.com/lattice/quda/commit/8bdf739483e267317bc4e7af7117b202907ecd35
https://github.com/lattice/quda/commit/ff3af36ad06315d7f9dea719009a3078fcbe85ce

maddyscientist commented on include/dirac_quda.h:
```diff
@@ -230,6 +230,16 @@ namespace quda {
+       @brief Similar to the Xpay version of Dslash, but used only by the Laplace op for 
+       smearing.
+    */    
+    virtual void SmearOp(ColorSpinorField &, const ColorSpinorField &, 
```
What makes this different from the `DslashXpay` that it necessitates its own function?

alexstrel replied:
It's an additional method, since requires a two-link field in addition to the regular staggered field. It's also optional, while dslash Xpay is mandatory.  (It is similar in terms of argument list , not the implementation, sorry for the confusion)

maddyscientist replied:
Ah thanks for the clarification @alexstrel.

mathiaswagner replied:
Please also clarify this in the doxygen comment.

maddyscientist commented on include/dslash_quda.h:
```diff
@@ -725,6 +725,17 @@ namespace quda
+     @param[in] dir Direction of the derivative 0,1,2,3 to omit (-1 is full 4D)
+     @param[in] a Scale factor applied to derivative
+     @param[in] b Scale factor applied to aux field
+  */
```
The doxygen doesn't match the list of parameters passed to the function.

maddyscientist commented on include/enum_quda.h:
```diff
@@ -571,6 +571,12 @@ typedef enum QudaGaugeSmearType_s {
   QUDA_GAUGE_SMEAR_INVALID = QUDA_INVALID_ENUM
 } QudaGaugeSmearType;
 
+typedef enum QudaFermionSmearType_s {
```
This addition also needs to be propagated to `enum_fortran_quda.h`

alexstrel replied:
done

maddyscientist commented on include/index_helper.cuh:
```diff
@@ -121,6 +121,19 @@ namespace quda {
+     Compute the checkerboard 1-d index from the 4-d coordinate x[] -2 in the mu direction
+
+     @return 1-d checkerboard index
+     @param x 4-d lattice index
```
Missing `[in]` / `[out]` labels from the parameters, and the `@return` should be after the parameter list.

maddyscientist commented on include/index_helper.cuh:
```diff
@@ -165,6 +178,20 @@ namespace quda {
+ /**
+     Compute the checkerboard 1-d index from the 4-d coordinate x[] +2 in the mu direction
+
+     @return 1-d checkerboard index
```
Missing [in] / [out] labels from the parameters, and the @return should be after the parameter list.

alexstrel replied:
fixed (also other places in the file)

maddyscientist commented on include/kernels/staggered_quark_smearing.cuh:
```diff
@@ -0,0 +1,229 @@
+{
+
+  /**
+     @brief Parameter structure for driving the covariatnt derivative operator
```
typo (covariatnt -> covariant)

alexstrel replied:
fixed

maddyscientist commented on include/kernels/staggered_quark_smearing.cuh:
```diff
@@ -0,0 +1,229 @@
+     @param[in] U The gauge field
+     @param[in] coord Site coordinate struct
+     @param[in] parity The site parity
+     @param[in] idx Thread index (equal to face index for exterior kernels)
```
Doxygen list doesn't quite match the function parameters being passed

alexstrel replied:
fixed

maddyscientist replied:
This still isn't quite fixed (`active` is missing)

maddyscientist commented on include/kernels/staggered_quark_smearing.cuh:
```diff
@@ -0,0 +1,229 @@
+            const Link U = arg.U(d, coord.x_cb, parity);
+            const Vector in = arg.in.Ghost(d, 1, ghost_idx, their_spinor_parity);//?
+
+            out += U * in;
```
I think this line, and equivalently on lines 127, 141 and 150 should be replaced with a call to `mv_add`.  This will ensure the operations are performed using the optimal number of `fma` instructions.  See for example [here](https://github.com/lattice/quda/blob/develop/include/kernels/dslash_staggered.cuh#L105).

alexstrel replied:
done (no difference in performance, though)

maddyscientist commented on include/kernels/staggered_quark_smearing.cuh:
```diff
@@ -0,0 +1,229 @@
+    static constexpr int nSpin  = 1;
+    static constexpr bool spin_project = false;
+    static constexpr bool spinor_direct_load = false; // false means texture load
+    typedef typename colorspinor_mapper<Float, nSpin, nColor, spin_project, spinor_direct_load>::type F;
```
style note: prefer the use of `using` over `typedef`.

alexstrel replied:
fixed

maddyscientist commented on include/kernels/staggered_two_link.cuh:
```diff
@@ -0,0 +1,102 @@
+
+        Link temp = Ucd*Ude;
+
+        arg.outA(mu, x_cb, parity) = temp;
```
you could jut write `arg.outA(mu, x_cb, parity) = Ucd * Ude;`

alexstrel replied:
sure, done.

maddyscientist commented on include/ks_qsmear.h:
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include <quda_internal.h>
+#include <quda.h>
```
Both the `quda.h` and `quda_internal.h` includes should be needed in this include.  We should be striving in general to only include what we use.

alexstrel replied:
removed `quda.h`

maddyscientist commented on include/quda.h:
```diff
@@ -1342,6 +1347,8 @@ extern "C" {
   void computeKSLinkQuda(void* fatlink, void* longlink, void* ulink, void* inlink,
                          double *path_coeff, QudaGaugeParam *param);
+                         
+  void computeTwoLinkQuda(void *twolink, void *inlink, QudaGaugeParam *param);
```
Add doxygen for this function.

alexstrel replied:
done

maddyscientist replied:
Missing `in/out` on `param` descriptor here

alexstrel replied:
fixed

maddyscientist commented on include/quda.h:
```diff
@@ -1701,6 +1708,21 @@ extern "C" {
+   *                     zero to keep two-link for future use (default = 0)
+   * @param t0          Set if the input spinor is on a time slice (default = -1)
+   */
+  void performTwoLinkGaussianSmearNStep(void *h_in, QudaInvertParam *inv_param, const int n_steps, const double width, const int compute_2link /*= 1*/, const int delete_2link /*= 0*/, const int t0 /*= -1*/);
```
Add missing `[in]` / `[out]` labels to this function.

maddyscientist replied:
The doxygen mentions that the parameters have default values, but the signature shows these are not present.  Indeed, default parameters are not allowed with regular C.

alexstrel replied:
as per our previous discussions, all params are now grouped in a new structure

maddyscientist replied:
This looks much cleaner now 😄

maddyscientist commented on CMakeLists.txt:
```diff
@@ -157,6 +157,8 @@ option(QUDA_COVDEV "build code for covariant derivative" OFF)
 option(QUDA_LAPLACE "build laplace operator" OFF)
 option(QUDA_QIO "build QIO code for binary I/O" OFF)
 
+option(QUDA_TWOLINK_GSMEAR "build code for two-link Gaussian smearing" OFF)
```
This option should have a name that will allow us to aggregate similar options together in the `ccmake` list.  E.g., `QUDA_SMEAR_GAUSS_TWOLINK` which would mean that if we have other `QUDA_SMEAR_....` options they will appear next to each other in the ccmake list.

alexstrel replied:
done

maddyscientist commented on lib/dirac_staggered.cpp:
```diff
@@ -138,6 +138,33 @@ namespace quda {
+      if (laplace3D == i) comm_dim[i] = 0;
+    }
+
+    ApplyStaggeredQSmear(out, in, *gauge, t0_local, is_time_slice, QUDA_INVALID_PARITY, laplace3D, dagger, comm_dim, profile); // parity is not used
```
`DiracStaggered::SmearOp` is missing the check for single parity that `DiracImprovedStaggered::SmearOp` has.

alexstrel replied:
done

maddyscientist commented on include/quda_milc_interface.h:
```diff
@@ -594,6 +594,17 @@ extern "C" {
+     Alias of qudaFreeGaugeSmeared().     
+     Free the two-link field allocated in QUDA.
+   */
+  void qudaFreeTwoLink();
```
Why introduce this alias?

alexstrel replied:
instead, deleted qudaFreeGaugeSmeared()

maddyscientist commented on tests/CMakeLists.txt:
```diff
@@ -110,6 +110,11 @@ if(QUDA_DIRAC_STAGGERED)
+  quda_checkbuildtest(staggered_gsmear_test QUDA_BUILD_ALL_TESTS)
+  install(TARGETS staggered_gsmear_test ${QUDA_EXCLUDE_FROM_INSTALL} DESTINATION ${CMAKE_INSTALL_BINDIR})
+
 endif()
```
There needs to be a `ctest` instance added in this file here as well so that the `staggered_gsmear_test` will run as part of regression tests.

maddyscientist commented on tests/utils/staggered_host_utils.cpp:
```diff
@@ -324,6 +326,150 @@ void computeLongLinkCPU(void **longlink, void **sitelink, QudaPrecision prec, vo
+
+void staggeredTwoLinkGaussianSmear(quda::ColorSpinorField &out, void *qdp_twolnk[], void** ghost_twolnk,  quda::ColorSpinorField &in, QudaGaugeParam* /*qudaGaugeParam*/, QudaInvertParam* /*inv_param*/, const int oddBit, const double /*width*/, const int /*t0*/, QudaPrecision prec)
+{
+  //const int nSrc = in.X(4);
```
Delete this line

alexstrel replied:
done

maddyscientist commented on tests/utils/host_utils.cpp:
```diff
@@ -906,6 +906,28 @@ int compare_floats(void *a, void *b, int len, double epsilon, QudaPrecision prec
+  return global_diff;
+}
+
+double compare_floats_v2(void *a, void *b, int len, double epsilon, QudaPrecision precision)
```
Why is a `v2` needed?  Add some comments to explain what `v2` does different from the original.

alexstrel replied:
it returns computed deviation rather then an int flag.

maddyscientist replied:
Ok, can you add a comment on this difference in the code then?

maddyscientist commented on tests/staggered_gsmear_test_utils.h:
```diff
@@ -0,0 +1,405 @@
+  QudaGaugeParam  gauge_param;//
+  QudaInvertParam inv_param;
+  //
+  std::unique_ptr<ColorSpinorField> spinor;
```
We never need to use smart pointers now for `ColorSpinorField` objects now that we have both default and move constructors.  These should be replaced with simply `ColorSpinorField` objects.

alexstrel replied:
not sure since it is in fact a conditional instantiation (e.g., the object is not needed for pure two-link tests)

maddyscientist replied:
That doesn't matter, since `ColorSpinorField` has a default constructor, so it's easy enough to conditionally create the object.

alexstrel replied:
done

maddyscientist commented on tests/staggered_gsmear_test_utils.h:
```diff
@@ -0,0 +1,405 @@
+
+    if (print_metrics) {
+      printfQuda("%fus per kernel call\n", 1e6 * gsmear_time.event_time / niter);
+#if 0
```
Is the code below going to enabled at some point?

alexstrel replied:
done

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -5356,6 +5443,160 @@ void performWuppertalnStep(void *h_out, void *h_in, QudaInvertParam *inv_param,
+  cudaParam.location = QUDA_CUDA_FIELD_LOCATION;
+  cudaParam.create   = QUDA_ZERO_FIELD_CREATE;
+  cudaParam.setPrecision(inv_param->cuda_prec, inv_param->cuda_prec, true);
+  ColorSpinorField *in    = ColorSpinorField::Create(cudaParam);
```
Can replace these with `ColorSpinorField in(cudaParam)`, etc.

alexstrel replied:
agreed, fixed

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -5356,6 +5443,160 @@ void performWuppertalnStep(void *h_out, void *h_in, QudaInvertParam *inv_param,
+    blas::zero(*temp1);
+  }
+
+#if 0
```
delete if not going to be used

alexstrel replied:
fixed

maddyscientist commented on lib/interface_quda.cpp:
```diff
@@ -5356,6 +5443,160 @@ void performWuppertalnStep(void *h_out, void *h_in, QudaInvertParam *inv_param,
 }
+ 
+
+void performTwoLinkGaussianSmearNStep(void *h_in, QudaInvertParam *inv_param, const int n_steps, const double width, const int compute_2link = 1, const int delete_2link = 0, const int t0 = -1)
```
It would be better to collect the parameters for this function into either a new struct , or to augment the `QudaInvertParam` with these new parameters.

maddyscientist replied:
Moreover, placing the parameters into the `QudaInvertParam` struct would allow for defaults to be set for these parameteres.

alexstrel replied:
done

maddyscientist commented on tests/staggered_gsmear_test_utils.h:
```diff
@@ -0,0 +1,405 @@
+    commDimPartitionedReset();
+  }
+
+  GSmearTime gsmearCUDA(int niter) //niter
```
Probably this function should not be called `gsmearCUDA` since QUDA runs on more than CUDA nowadays 😉

alexstrel replied:
sure, done.

maddyscientist commented on lib/staggered_quark_smearing.cu:
```diff
@@ -0,0 +1,221 @@
+
+      // operator is Hermitian so do not instantiate dagger
+      if (arg.nParity == 1) {
+        if(arg.is_t0_kernel) //FIXME we reuse xpay non-type template parameter for t0 option!
```
Is this really a `FIXME` or just helpful information?  I think abusing the `xpay` template like this seems like a reasonable thing to do.

maddyscientist replied:
Actually thinking about this a bit more, I think we might want to fix this, so that it doesn't use the `xpay` template.  

(The `xpay` template is something that I think I might want to explore removing in the future.  For example, making it a run-time option when fast-compilation for the dslash is enabled).

For the present case here, maybe just make the `is_t0_kernel` parameter a run-time variable.  Looking at the code, it doesn't look like this really needs to be a compile-time parameter and run-time should be sufficient.  This will of course also reduce compilation time.

alexstrel replied:
fixed

maddyscientist commented on lib/staggered_two_link_quda.cu:
```diff
@@ -0,0 +1,74 @@
+      checkLocation(newTwoLink, link);
+      checkPrecision(newTwoLink, link);
+      instantiate<ComputeTwoLink, WilsonReconstruct>(link, newTwoLink);
+
```
delete lines 65-66 (general style it to leave trivial `return` statements as implicit)

alexstrel replied:
done

maddyscientist reviewed (changes requested):
I've left initial review comments.  This is really great work @alexstrel, @hwancheolJeong and @stevengottlieb, and is a great example of the Dslash framework in action 😄   Thank you for this.

The main things to be addressed are documentation, cleaning up the doxygen comments, some adaptions for modern QUDA (e.g., just use `ColorSpinorField` objects instead of (smart) pointers), and finally getting some tests added to the `ctest` list.

alexstrel committed:
first update to address PR comments

alexstrel committed:
a few corrections to address PR comments

alexstrel committed:
reordered smearing params in a single structure

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

alexstrel committed:
fixed compilation errors

hwancheolJeong committed:
Fix mv_add usage and remove extra \n from errorQuda.

hwancheolJeong committed:
minor fixes

hwancheolJeong committed:
Remove the use of xpay

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

alexstrel committed:
updated flops count

alexstrel committed:
a few changes to fix flops reports

alexstrel committed:
added comment

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

alexstrel committed:
fixed gauge field load method (now loaded as wilson links)

alexstrel committed:
replaced color spinor pointers by objects

alexstrel committed:
a number of minor fixes

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

maddyscientist commented on tests/utils/command_line_params.cpp:
```diff
@@ -272,6 +272,12 @@ double eofa_mq3 = 1.0;
 
 QudaContractType contract_type = QUDA_CONTRACT_TYPE_OPEN;
 
+int n_steps = 50;
```
Can you put all these variables in a struct, so that it's clear what option group they belong to?

alexstrel replied:
Thanks @maddyscientist  for the review. Not sure I understand the last point. Would you like me to create a common structure that will include all related variables? Currently I just left a comment above the group while all variables in the group remain plain build-in types. (Similar to other test vars, e.g. deflation-related ones)

maddyscientist replied:
How about just prefixing these variables, e.g., `smear_n_steps`, `smear_t0`, etc. so it's clear which group they belong to?

alexstrel replied:
sure, done.

maddyscientist replied:
[35d2002](https://github.com/lattice/quda/pull/1312/commits/35d2002fcc6ce30ed86e8c59e8cd517365815db1)

maddyscientist commented:
Thanks @alexstrel for the updates to the PR.  I left some more minor comments.  Beyond these, the main outstanding issue is the lack of including the test in the `ctest` list.

alexstrel committed:
added quark smearing in the ctest list

alexstrel committed:
corrected doxygen for the smearing routines

alexstrel committed:
added prefixes for smearing variables

weinbe2 commented:
@alexstrel do you have a MILC workflow leveraging this that you can share? It can be a dummy one, no need to share anything special. As part of testing on my end it'd just be nice to see it "in action".

mathiaswagner commented on tests/host_reference/gsmear_test_helpers.h:
```diff
@@ -0,0 +1,10 @@
+#pragma once
```
Do we need that as a separate header?

alexstrel replied:
fixed, enumerator is moved into a destination header

mathiaswagner commented on lib/dirac_improved_staggered.cpp:
```diff
@@ -127,6 +127,37 @@ namespace quda {
     // do nothing
   }
+  
+  void DiracImprovedStaggered::SmearOp(ColorSpinorField &out, const ColorSpinorField &in, const double &, const double &,
```
is this an exact code duplication from `dirac_staggered.cpp` ? Just wondering if we can avoid that.

hwancheolJeong replied:
They are identical. To avoid duplication, I think we should introduce an intermediate class between the Dirac class and them (DiracStaggered and DiracImprovedStaggered). Or, if we remove (or replace) checkSpinorAlias() defined in the Dirac class, it would also be possible to create a shared function outside the class (or in another mother class).

maddyscientist replied:
I think we can leave this as it is for now and consider cleanup / consolidation after merging.

mathiaswagner commented on lib/CMakeLists.txt:
```diff
@@ -287,6 +287,22 @@ if(QUDA_INTERFACE_TIFR OR QUDA_INTERFACE_ALL)
   target_compile_definitions(quda PUBLIC BUILD_TIFR_INTERFACE)
 endif(QUDA_INTERFACE_TIFR OR QUDA_INTERFACE_ALL)
 
+if(QUDA_CONTRACT)
```
looks like the target_compile_definitions should not be added back. IIRC @maddyscientist removed those in favor of a configured header.

maddyscientist replied:
Ah, good catch @mathiaswagner

hwancheolJeong replied:
If I understand correctly, it's not just a problem for this pr to fix but a general issue. Right?

maddyscientist replied:
Resolved with [daa6244](https://github.com/lattice/quda/pull/1312/commits/daa6244e041e55034a1adbe8cc0a47bd86b790b6)

mathiaswagner commented on include/dirac_quda.h:
```diff
@@ -1361,6 +1371,18 @@ namespace quda {
+     * @param[in] t0 time-slice index
+     * @param[in] parity Parity flag
+     */
+    void SmearOp(ColorSpinorField &out, const ColorSpinorField &in, const double &a, const double &b, const int &t0,  const QudaParity parity) const;
```
Why do we pass doubles as const ref instead of by value?

hwancheolJeong replied:
Replied at https://github.com/lattice/quda/pull/1312#discussion_r1025695761

mathiaswagner commented on lib/interface_quda.cpp:
```diff
@@ -3932,6 +3945,80 @@ void computeKSLinkQuda(void *fatlink, void *longlink, void *ulink, void *inlink,
+
+  profileGaussianSmear.TPSTART(QUDA_PROFILE_FREE);
+  
+  delete gaugeSmeared;gaugeSmeared = nullptr;
```
that should not be a single line but I think we need a clang-format on the diff anyway

hwancheolJeong replied:
Thanks for finding it out. Split into two lines. https://github.com/lattice/quda/commit/8e7ad58492d3a863a6af6f868dd28d5354de57e8

alexstrel committed:
fixed CMakeList.txt file  (added missed prefix)

alexstrel committed:
fixed prefix

hwancheolJeong commented:
> @alexstrel do you have a MILC workflow leveraging this that you can share? It can be a dummy one, no need to share anything special. As part of testing on my end it'd just be nice to see it "in action".

@weinbe2 Here is my test app: https://github.com/hwancheolJeong/milc_qcd/tree/feature/gaussianSmearing/ks_gauss_smear
It is my fork of MILC code. It uses an interface defined in generic_ks/gauss_smear_ks_QUDA.c that will be merged into develop branch later (once this pr is done.), but this test app is solely for personal use. 
You can build the test app by "make ks_gauss_smear", and run it with an input file (refer to test/ks_gauss_smear.2.sample-in). It will produce a smeared source in Scidac format as well as binary format (for easy comparison).

hwancheolJeong committed:
Split two executions into two lines.

hwancheolJeong committed:
Change a,b,t0 parameters of SmearOP() to be called by value

hwancheolJeong committed:
fix doxygen description

alexstrel committed:
corrected README (added Hwancheol into the author list)

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

alexstrel committed:
removed target_compile_definitions (moved into the configure header)

maddyscientist commented:
Thanks for the latest updates @alexstrel and @hwancheolJeong.  I've done some tests with the latest version, and things are mostly good.  A few outstanding issues before we can merge:
* the automatic ctest only runs in single precision, can you add a double precision version as well?  (see what is done with `gauge_path_test` for example for how to deal with this, where it is run in a loop over enabled precision)
* reconstruct options seem to be ignored by the test, and the kernel is always run with no reconstruct
* can you merge in #1345 into your PR?  This will be merged shortly into develop, but in the meantime it will fix the presently broken Jenkins build with CUDA 12 / GCC-12 / Clang 14

There's some other cleanup that I will apply, but this will be for after the branch is merged.

maddyscientist commented:
....and lastly, this PR needs a clang-format.

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

alexstrel committed:
added double precision test

alexstrel committed:
applied clang-format

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

alexstrel committed:
resctricted reconstruction options to RECONSTRUCT_NO

alexstrel committed:
fixed missing in in param descriptor
removed gsmear_test_helper.h
replaced smart pointers for ColorSpinorField objects

alexstrel committed:
fixed compiler errors due to previous commit

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

maddyscientist reviewed (approved):
Thank you @alexstrel and @hwancheolJeong for continuing to get this PR in good shape.  I'm finally approving 😉 

@weinbe2 can you take this PR for a spin before we merge?

weinbe2 committed:
Small doxygen updates

weinbe2 committed:
Enforce that the staggered dslash must be enabled to compile two link gaussian smearing support

weinbe2 commented:
@alexstrel and @hwancheolJeong , this is all really nice! I'm doing some testing and a little bit of cleanup here and there, but this should be good to go soon.

I had one question --- I noticed that you created a routine in `milc_interface.cpp`, exposed in the appropriate header, to free the two link gauge field, but no routines to create the field or apply the smearing. Is this intentional?

weinbe2 commented on include/kernels/staggered_quark_smearing.cuh:
```diff
@@ -0,0 +1,222 @@
+      int thread_dim;
+
+      if (arg.is_t0_kernel) {
+        if (arg.t0 < 0) return;
```
I'm worried that this `return;` could imply that `arg.out(...);` is never assigned, which may happen if this is a case where line 218:
```
if (kernel_type != EXTERIOR_KERNEL_ALL || active) arg.out(coord.x_cb, my_spinor_parity) = out;
```
never gets hit.

I think we should refactor the logic here in a way to at least guarantee that `arg.out` always has something sane assigned to it.

hwancheolJeong replied:
Regarding your first concern, I am afraid I don't know much about other backends. What can happen on them due to `return;`?

Regarding your second concern, if I understand correctly and my memory is correct, I recall other kernels are defined in a similar way. (e.g., https://github.com/lattice/quda/blob/develop/include/kernels/dslash_staggered.cuh#L206)
And arg.out is assigned here: https://github.com/lattice/quda/blob/feature/quark-smearing.pr/include/kernels/staggered_quark_smearing.cuh#L50.

hwancheolJeong replied:
For the latter one, are you concerned when ColorSpinorField instance `out` is created with NULL_FIELD_CREATE option? It can happen in general.. though we are creating `out` with ZERO_FIELD_CREATE in the `performTwoLinkGaussianSmearNStep()`..

hwancheolJeong replied:
And I believe I checked all available dslash policies... (nothing in memory though..)

maddyscientist replied:
@hwancheolJeong I think @weinbe2's concern with other targets is misplaced, and there's no issue there.  And I agree with you on the second point as well, there's no issue with `arg.out` not being hit here, and this is consistent with what happens in the other dslash kernels.  Marking as resolved.

weinbe2 commented on lib/staggered_quark_smearing.cu:
```diff
@@ -0,0 +1,223 @@
+
+    void apply(const qudaStream_t &stream)
+    {
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
```
While this isn't going to break correctness, when `t0 == -1` on rank 0 the tuned values for the dslash part of this are going to end up very meaningless, since the kernel isn't doing any work...

hwancheolJeong replied:
Thanks. I (or we) didn't think about it... Now, I am concerned.. What would happen when the second source's time slice is on a different rank from the first source's?

maddyscientist replied:
I just had a chance to look into this, and I share @weinbe2's concern here.  I'm not sure what the solution is to this.  Perhaps we could have an interface where we can override which rank does the kernel tuning and broadcasts the tune cache from?  Then here we'd set this to being the first rank with the time slice in it.

weinbe2 commented on tests/staggered_gsmear_test.cpp:
```diff
@@ -0,0 +1,87 @@
+
+  display_test_info();
+
+  if (link_recon != QUDA_RECONSTRUCT_NO) {
```
Out of curiosity, what's the reason for enforcing this, for at least `computeTwoLink`? As far as I can tell, `computeTwoLink` supports reconstruct 12 and 8 (and you instantiate it).

I can see how it wouldn't work for `staggeredTwoLinkGaussianSmear`, though.

alexstrel replied:
Reconstruction options are left for the future optimization work.

hwancheolJeong commented:
> Thank you @alexstrel and @hwancheolJeong for continuing to get this PR in good shape. I'm finally approving 😉
> 
> @weinbe2 can you take this PR for a spin before we merge?

@maddyscientist I appreciate you! for reviewing it and for your advice and help.

hwancheolJeong commented:
> @alexstrel and @hwancheolJeong , this is all really nice! I'm doing some testing and a little bit of cleanup here and there, but this should be good to go soon.
> 
> I had one question --- I noticed that you created a routine in `milc_interface.cpp`, exposed in the appropriate header, to free the two link gauge field, but no routines to create the field or apply the smearing. Is this intentional?

Thanks, @weinbe2 .
I made an interface inside the MILC code: https://github.com/hwancheolJeong/milc_qcd/blob/feature/gaussianSmearing/generic_ks/gauss_smear_ks_QUDA.c,
which will be merged into the develop branch of the MILC code once this PR is done (I may need to update a little to accommodate recent updates in QUDA).
We may add a similar interface routine in milc_interface.cpp later.

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

maddyscientist commented on lib/staggered_two_link_quda.cu:
```diff
@@ -0,0 +1,67 @@
+    checkNative(newTwoLink, link);
+    checkLocation(newTwoLink, link);
+    checkPrecision(newTwoLink, link);
+    instantiate<ComputeTwoLink, WilsonReconstruct>(link, newTwoLink);
```
Should this be `WilsonReconstruct` here or is `StaggeredReconstruct` more appropriate?

weinbe2 replied:
I dug into this and realized `WilsonReconstruct` does make sense at face value, since the smearing does indeed seem to be done with the non-phased links. @hwancheolJeong I did see you have a comment in your test executable and your MILC code that `recon-12` isn't working, though. If that's the case, can you replace the `WilsonReconstruct` with `ReconstructNo` (so it'll be recon-18 only) and file an issue for it in this repo so it's revisited in the future?

alexstrel replied:
thanks @weinbe2 I agree it makes sense to restrict the reconstruction here as well , (with a fixme note)

alexstrel committed:
Merge branch 'develop' into feature/quark-smearing.pr

weinbe2 commented:
> > @alexstrel and @hwancheolJeong , this is all really nice! I'm doing some testing and a little bit of cleanup here and there, but this should be good to go soon.
> > I had one question --- I noticed that you created a routine in `milc_interface.cpp`, exposed in the appropriate header, to free the two link gauge field, but no routines to create the field or apply the smearing. Is this intentional?
> 
> Thanks, @weinbe2 . I made an interface inside the MILC code: https://github.com/hwancheolJeong/milc_qcd/blob/feature/gaussianSmearing/generic_ks/gauss_smear_ks_QUDA.c, which will be merged into the develop branch of the MILC code once this PR is done (I may need to update a little to accommodate recent updates in QUDA). We may add a similar interface routine in milc_interface.cpp later.

Understood, thank you for clarifying. That's not the typical approach but I'm certainly not going to hold up the PR on this! Similar to my other ask, can you please file an issue noting that this needs to be ported into the MILC Interface in the future?

alexstrel committed:
disabled gauge link reconstruction in the two-link computation. This is
already disabled in the high level interface functions.

alexstrel committed:
fixed compilation errors due to last commit

weinbe2 reviewed (approved):
After some offline discussion, I'd say this is ready to go, and the outstanding questions (which are ultimately harmless for this PR's correctness/functionality) can be addressed in a future PR. Awesome work @alexstrel and @hwancheolJeong !

alexstrel committed:
fixed tunekey arguments in laplace.cu and covDev.cu

weinbe2 commented:
I'll merge this after Jenkins completes
